Redesign nav bar, GlobalSearch, and dropdowns

taneliang opened the pull request:
<!--
Thank you for contributing to NUSMods!
The template below was made to help both you and the reviewers understand
your changes. Please fill it up to the best of your ability.
(These are comments, they won't be shown in the "preview")
-->

## Summary

This PR redesigns our nav bar, global search, and all dropdowns throughout the app.

The design is inspired by FB5 (the recent facebook.com redesign) and @jiholim's NUSMods V4 mockups from 2018.

Because a huge number of things have been changed, it'll probably be easier to review this PR without looking at the existing code or behavior.

Fixes #3060 and resolves #1646 (no sidebar => no problem)

Thanks to @eugenood, @williamhutech, @chrisgzf, and @chuabingquan for the valuable input!

### Redesigned nav bar

#### Why

We need space to put a lot more buttons in the future, including user account/info, and possibly an MPE button and planner button. The tab bar is full on mobile, and the top bar is full on desktop, leaving no good place to put auth info. This PR unifies both the mobile and desktop nav experience, and provides space on the right for logged-in user info + a dropdown menu for extra pages e.g. Contribute, Settings, MPE, etc.

#### Design Inspiration

FB5:

![image](https://user-images.githubusercontent.com/12784593/105956391-b1e5ba80-60b2-11eb-8b62-5c155dab80b8.png)

@jiholim's mockup:

![image](https://user-images.githubusercontent.com/12784593/105956427-bdd17c80-60b2-11eb-8aec-8c88712f92a8.png)

GitHub:

![image](https://user-images.githubusercontent.com/12784593/105957044-a1820f80-60b3-11eb-9875-352b74d740a8.png)

#### What Changed

- Nav tabs on `sm`-width and above are moved to the nav bar.
- A nav dropdown has been added.
- AY week string has been moved into the dropdown, and an academic calendar link has been added.
- NUS Business link added.
- Global search is now always expanded, for simpler code and greater visibility.
- Centered page content, and tweak Today and Venues page layouts.

| |Before|After|
|-|-|-|
|320px (`xs`)|![nusmods com_modules_CS1010E_programming-methodology](https://user-images.githubusercontent.com/12784593/105959727-43572b80-60b7-11eb-838b-3f0c809e782e.png)|![nusmods-website-git-eliang-new-nav nusmodifications vercel app_modules_CS1010E_programming-methodology](https://user-images.githubusercontent.com/12784593/105959823-5e29a000-60b7-11eb-9ba2-ed6c3b48ad42.png)|
|1024px (`sm`)|![nusmods com_modules_CS1010E_programming-methodology (1)](https://user-images.githubusercontent.com/12784593/105960087-b496de80-60b7-11eb-98ad-6aa7a7542ba1.png)|![nusmods-website-git-eliang-new-nav nusmodifications vercel app_modules_CS1010E_programming-methodology (2)](https://user-images.githubusercontent.com/12784593/105960114-bc568300-60b7-11eb-8d6c-f41fe93a07fb.png)|
|1440px (`xl`)|![nusmods com_modules_CS1010E_programming-methodology (2)](https://user-images.githubusercontent.com/12784593/105960238-e60faa00-60b7-11eb-9599-31ce712bcf01.png)|![nusmods-website-git-eliang-new-nav nusmodifications vercel app_modules_CS1010E_programming-methodology (3)](https://user-images.githubusercontent.com/12784593/105960243-e8720400-60b7-11eb-9b5a-dc7aee05a043.png)|

Nav dropdown on xs

![nusmods-website-git-eliang-new-nav nusmodifications vercel app_modules_CS1010E_programming-methodology (1)](https://user-images.githubusercontent.com/12784593/105959931-81ece600-60b7-11eb-9be1-0a4e21d653e4.png)

### Redesigned dropdown

#### Why

The existing dropdown did not look particularly good, and would have been weird as a nav dropdown. As I had designed a new one for the nav dropdown based on FB5's design (but it's also pretty common elsewhere), I just moved that into our global styles for consistency throughout the whole app.

The new dropdown styles also include CSS classes for items with icons and right contents.

#### Design Inspiration

FB5

![image](https://user-images.githubusercontent.com/12784593/105956815-52d47580-60b3-11eb-987d-95fea7cc4bd0.png)

#### Screenshots

| |Before|After|
|-|-|-|
|`ExportMenu`|![image](https://user-images.githubusercontent.com/12784593/105957435-384ecc00-60b4-11eb-8e55-d84c42aecf2b.png)|![image](https://user-images.githubusercontent.com/12784593/105957493-4dc3f600-60b4-11eb-8d74-bba81c45cf3b.png)|
|`DropdownListFilters`|![image](https://user-images.githubusercontent.com/12784593/105957602-70eea580-60b4-11eb-9ebf-b7270e00b783.png)|![image](https://user-images.githubusercontent.com/12784593/105957583-6a602e00-60b4-11eb-9a30-d739a2359ba8.png)|
|`AddModuleDropdown`|![image](https://user-images.githubusercontent.com/12784593/105957784-b01cf680-60b4-11eb-84db-539c617732df.png)|![image](https://user-images.githubusercontent.com/12784593/105957744-a1ceda80-60b4-11eb-9ab8-247d98e6a141.png)|

### Redesigned global search

#### Why

Our `GlobalSearch` design was a little inconsistent with the rest of NUSMods, and was extremely awkward when used with the new nav bar design, and felt dated and out of place when placed beside the new dropdown.

#### Design Inspiration

@jiholim's mockup:

![image](https://user-images.githubusercontent.com/12784593/105956077-371c9f80-60b2-11eb-860e-e4114b744fa0.png)

#### What Changed

- Dropdown now uses the global `dropdown-menu` styles instead of its own custom stuff.
- The section headers (i.e. Modules and Venues) are now sticky, though this is broken on iOS. I don't know why sticky is broken on iOS, but we can fix that in a future PR as the existing design doesn't have sticky at all so we've already improved on it 😆.
- The section headers are no longer clickable; only the View All buttons are.
- The dropdown no longer expands as wide as it used to.

| |Before|After|
|-|-|-|
|Both modules and venues present|![image](https://user-images.githubusercontent.com/12784593/105958753-00e11f00-60b6-11eb-876b-57ffb918ea88.png)|![image](https://user-images.githubusercontent.com/12784593/105958731-fb83d480-60b5-11eb-9883-17951cc84c0c.png)|
|Only modules|![image](https://user-images.githubusercontent.com/12784593/105958888-2cfca000-60b6-11eb-9ef2-8541a7ed8b6e.png)|![image](https://user-images.githubusercontent.com/12784593/105958903-3554db00-60b6-11eb-83dc-685210ac9430.png)|
|View All hover|![image](https://user-images.githubusercontent.com/12784593/105959049-5f0e0200-60b6-11eb-962b-4882c1060796.png)|![image](https://user-images.githubusercontent.com/12784593/105959032-57e6f400-60b6-11eb-8fff-c8ed8b9cf4aa.png)|
|`sm` viewport|![image](https://user-images.githubusercontent.com/12784593/105959234-9e3c5300-60b6-11eb-995d-73198da94019.png)|![image](https://user-images.githubusercontent.com/12784593/105959190-8d8bdd00-60b6-11eb-8974-025da530daa1.png)|

taneliang committed:
WIP move navtabs into navbar

taneliang committed:
Don't left align content anymore

taneliang committed:
Allow Venues page map to expand

taneliang committed:
Make nav bar white with bottom border

White background allows more colors to be used in the nav bar.

Feels like there are too many horizontal lines in the app, but just
playing around for now. We can try fixing this later.

taneliang committed:
Add bottom line to navtabs

taneliang committed:
Replace navbar bottom border with softer shadow

taneliang committed:
Roughly allow GlobalSearch to expand rightwards

taneliang committed:
Implement expanding GlobalSearch

chrisgzf commented:
I like the idea of this! Could possibly be a fresh new change with some tweaks! How many buttons are you expecting for the mobile version of this. Might be a little cramped (espc for 320px viewports).

chrisgzf commented:
Toyed around with smth like this for Hack&Roll. For 1 row-nav, it felt a bit cramped already.

<img width="353" alt="image" src="https://user-images.githubusercontent.com/4933577/104733734-1d0dc380-577a-11eb-8e78-551c67c2b736.jpg">

On 320px
<img width="353" alt="image" src="https://user-images.githubusercontent.com/4933577/104733787-2bf47600-577a-11eb-9a32-767e30a75102.png">

Are we sticking with the 2-row nav like we currently have for mobile?

taneliang committed:
Merge branch 'master' into eliang/new-nav

taneliang committed:
Move GlobalSearch to the right

taneliang committed:
Replace top/bottomBar with flex layouts

bottomBar was position:relative, which caused it to float above the page
regardless of where its parent nav bar was. This causes the tabs to
stick to the top of the screen when you rubberband-scroll on iOS.

To fix this, this commit brings the tabs inline, laying out the nav bar
using flex magic instead.

This also allows us to easily prevent the search box or other nav
bar components from overlapping with the tabs. Global search is now much
less buggy than before.

taneliang committed:
Enlarge logo tappable area

taneliang committed:
Remove $side-nav-width-* variables

taneliang committed:
Add transition to search box open/close

taneliang committed:
Merge branch 'master' into eliang/new-nav

taneliang committed:
Only animate opacity for GlobalSearch input to prevent icon jump at end of exit

taneliang committed:
WIP nav dropdown

taneliang commented:
Hey @chrisgzf! Sorry I didn't notice that you commented on this haha.

> For 1 row-nav, it felt a bit cramped already.

Yeah we're definitely not gonna do a 1-row nav on mobile. I'm doing something similar to the current 2-row nav.

![image](https://user-images.githubusercontent.com/12784593/104808414-18e8b100-5821-11eb-9ffa-27fd16e0bf38.png)

taneliang committed:
Merge branch 'master' into eliang/new-nav

taneliang committed:
Implement NavDropdown

taneliang committed:
Add accessory info to dropdown items

taneliang committed:
Remove background color from circles

taneliang committed:
Shorten search placeholder to fit

taneliang committed:
Move global search dropdown up a little

taneliang committed:
Fix nav circle and GlobalSearch colors

taneliang committed:
Enable Navtabs text on lg

taneliang committed:
Fix dropdown divider in planner button

taneliang committed:
Fix NavDropdown toggle, hover states

taneliang committed:
Add NUS Business, rename NUSWhispers, improve right accessory icon color

taneliang committed:
Replace chevron button with hamburder

taneliang committed:
Merge branch 'master' into eliang/new-nav

taneliang committed:
Merge branch 'master' into eliang/new-nav

taneliang committed:
Make all nav buttons color: var(--gray-mid)

taneliang committed:
Make GlobalSearch always opened

Single commit so that this can hopefully be reverted easily if we prefer
collapsing.

taneliang committed:
Use Downshift's isOpen in GlobalSearch

taneliang committed:
Give up on avoiding Navbar shadow appearing over SearchkitSearchBox

taneliang committed:
Remove $navtabs-z-index

taneliang committed:
Clean up code

taneliang committed:
stylelint --fix

taneliang committed:
Update Navtabs -> Navbar tests

taneliang committed:
Add AY/sem/week text to dropdown

taneliang committed:
Add AY calendar links to dropdown

taneliang committed:
Merge branch 'master' into eliang/new-nav

taneliang committed:
Misc NavDropdown fixes

taneliang committed:
Merge branch 'master' into eliang/new-nav

taneliang committed:
Redesign global dropdown using NavDropdown styles

taneliang committed:
Fix DropdownListFilters funky checkboxes with new dropdown-items

taneliang committed:
Move dropdown-item with icons styles into global styles

taneliang committed:
Redesign GlobalSearch

taneliang committed:
Fix AddModuleDropdown spacing

taneliang committed:
Move SemesterBadge margin to its left side

taneliang committed:
Remove unnecessary styles

taneliang committed:
Delete unused AppShell.scss

taneliang committed:
Remove width:100% from global dropdown-menu to fix ExportMenu truncation

taneliang committed:
GlobalSearch: allow focus to open dropdown

taneliang committed:
GlobalSearch: Add aria-label to view all buttons

taneliang committed:
Update Navbar test snapshots

taneliang committed:
Merge branch 'master' into eliang/new-nav

bnjmnt4n commented:
I quite like the look of the new components, and the removal of the sidebar, but one area that bugged me was the centered links in the navbar, especially on larger viewports. I think there's a distinct feeling of the navbar being unbalanced, when the links are in the center and the large-ish search field on the right. I think a good solution would simply be to shift the navigation links to the left, next to the NUSMods logo. Will send a PR with screenshots later for reference/comparison.
